File: js/drawing/controller.js
import * as THREE from 'three';

/**
 * Simple 1D Kalman filter for smoothing noisy measurements
 */
class KalmanFilter1D {
    /**
     * @param {number} Q - Process noise (lower = smoother, slower response)
     * @param {number} R - Measurement noise (higher = smoother, trusts measurements less)
     */
    constructor(Q = 0.1, R = 0.5) {
        this.Q = Q; // Process noise
        this.R = R; // Measurement noise
        this.x = 0; // Estimated value
        this.P = 1; // Estimation error covariance
        this.initialized = false;
    }

    /**
     * Update filter with new measurement
     * @param {number} measurement - The raw measurement
     * @returns {number} - The filtered value
     */
    update(measurement) {
        if (!this.initialized) {
            this.x = measurement;
            this.initialized = true;
            return this.x;
        }

        // Prediction
        this.P = this.P + this.Q;

        // Update
        const K = this.P / (this.P + this.R); // Kalman gain
        this.x = this.x + K * (measurement - this.x);
        this.P = (1 - K) * this.P;

        return this.x;
    }

    reset() {
        this.x = 0;
        this.P = 1;
        this.initialized = false;
    }
}

export class Controller extends THREE.Object3D {
    /**
     * @param {Object} [kalmanConfig] - Kalman filter configuration
     * @param {number} [kalmanConfig.Q=0.1] - Process noise (lower = smoother)
     * @param {number} [kalmanConfig.R=0.5] - Measurement noise (higher = smoother)
     * @param {boolean} [kalmanConfig.enabled=true] - Whether filtering is enabled
     */
    constructor(kalmanConfig = {}) {
        super();
        // Grip button (Closed_Fist to engage, Open_Palm to release)
        this.grip_Down = false;
        this.grip_Held = false;
        this._wasGrip_Held = false;

        // Trigger button
        this.trigger_Down = false;
        this.trigger_Held = false;
        this._wasTrigger_Held = false;

        // Button A
        this.buttonA_Down = false;
        this.buttonA_Held = false;
        this._wasButtonA_Held = false;

        // Button B
        this.buttonB_Down = false;
        this.buttonB_Held = false;
        this._wasButtonB_Held = false;

        // Button C
        this.buttonC_Down = false;
        this.buttonC_Held = false;
        this._wasButtonC_Held = false;

        // Grip timeout - release grip if closed_fist not seen for this duration
        this.gripTimeout = 1000; // ms
        this._lastClosedFistTime = 0;

        // Kalman filter setup - separate filters for drawing and navigation
        // Base smoothing: Q=0.1, R=0.5
        // Drawing: 50% smoothing (less smooth, more responsive)
        // Navigation: 200% smoothing (more smooth, less responsive)
        const baseQ = kalmanConfig.Q ?? 0.1;
        const baseR = kalmanConfig.R ?? 0.5;
        this.kalmanEnabled = kalmanConfig.enabled ?? true;

        // Drawing filters (50% = half the smoothing, double Q, halve R)
        const drawQ = baseQ * 2;
        const drawR = baseR * 0.5;
        this._drawKalmanX = new KalmanFilter1D(drawQ, drawR);
        this._drawKalmanY = new KalmanFilter1D(drawQ, drawR);
        this._drawKalmanZ = new KalmanFilter1D(drawQ, drawR);
        this._drawPosition = new THREE.Vector3();

        // Navigation filters (200% = double the smoothing, halve Q, double R)
        const navQ = baseQ * 0.5;
        const navR = baseR * 2;
        this._navKalmanX = new KalmanFilter1D(navQ, navR);
        this._navKalmanY = new KalmanFilter1D(navQ, navR);
        this._navKalmanZ = new KalmanFilter1D(navQ, navR);
        this._navPosition = new THREE.Vector3();

        // Confidence scoring
        this.confidence = 1.0;
        this._prevPosition = new THREE.Vector3();
        this._prevVelocity = new THREE.Vector3();
        this._hasHistory = false;
        this._confidenceSmoothing = 0.1; // EMA factor (lower = smoother)
        this._buttonsBlocked = false;
        this._blockThreshold = 0.2;
        this._unblockThreshold = 0.4;
        // Tuning: acceleration magnitude that maps to confidence ~0.5
        this._accelMidpoint = 0.15;
    }

    /**
     * Updates the rolling confidence score based on motion smoothness.
     * Uses acceleration magnitude as a measure of jerkiness.
     * @private
     * @param {THREE.Vector3} currentPosition
     */
    _updateConfidence(currentPosition) {
        if (!this._hasHistory) {
            this._prevPosition.copy(currentPosition);
            this._prevVelocity.set(0, 0, 0);
            this._hasHistory = true;
            return;
        }

        // Calculate current velocity
        const velocity = new THREE.Vector3().subVectors(currentPosition, this._prevPosition);

        // Calculate acceleration (change in velocity)
        const acceleration = new THREE.Vector3().subVectors(velocity, this._prevVelocity);
        const accelMag = acceleration.length();

        // Map acceleration to instantaneous confidence
        // accelMag = 0 -> confidence = 1.0
        // accelMag = _accelMidpoint -> confidence ~= 0.5
        // accelMag = high -> confidence -> 0
        const instantConfidence = Math.exp(-accelMag / this._accelMidpoint * 0.693); // 0.693 = ln(2)

        // Exponential moving average for smooth rolling score
        this.confidence = this.confidence * (1 - this._confidenceSmoothing)
                        + instantConfidence * this._confidenceSmoothing;
        this.confidence = Math.max(0, Math.min(1, this.confidence));

        // Hysteresis for button blocking
        if (this._buttonsBlocked) {
            if (this.confidence > this._unblockThreshold) {
                this._buttonsBlocked = false;
            }
        } else {
            if (this.confidence < this._blockThreshold) {
                this._buttonsBlocked = true;
            }
        }

        // Store for next frame
        this._prevPosition.copy(currentPosition);
        this._prevVelocity.copy(velocity);
    }

    /**
     * Gets the drawing position (50% smoothing - more responsive)
     * @param {THREE.Vector3} target - Vector to store the result
     * @returns {THREE.Vector3} The drawing position in world space
     */
    getDrawPosition(target) {
        if (!target) target = new THREE.Vector3();
        // Transform local draw position to world space
        target.copy(this._drawPosition);
        if (this.parent) {
            this.parent.localToWorld(target);
        }
        return target;
    }

    /**
     * Gets the navigation position (200% smoothing - smoother)
     * @param {THREE.Vector3} target - Vector to store the result
     * @returns {THREE.Vector3} The navigation position in world space
     */
    getNavPosition(target) {
        if (!target) target = new THREE.Vector3();
        target.copy(this._navPosition);
        if (this.parent) {
            this.parent.localToWorld(target);
        }
        return target;
    }

    /**
     * Updates only the controller's pose (position and rotation).
     * @param {THREE.Vector3} position - The new position of the controller
     * @param {THREE.Quaternion} [rotation] - The new rotation of the controller (optional)
     */
    updatePose(position, rotation) {
        if (position) {
            if (this.kalmanEnabled) {
                // Update drawing position (50% smoothing - more responsive)
                this._drawPosition.set(
                    this._drawKalmanX.update(position.x),
                    this._drawKalmanY.update(position.y),
                    this._drawKalmanZ.update(position.z)
                );

                // Update navigation position (200% smoothing - smoother)
                this._navPosition.set(
                    this._navKalmanX.update(position.x),
                    this._navKalmanY.update(position.y),
                    this._navKalmanZ.update(position.z)
                );

                // Use navigation position for the controller's base position
                this.position.copy(this._navPosition);
            } else {
                this.position.copy(position);
                this._drawPosition.copy(position);
                this._navPosition.copy(position);
            }

            // Calculate confidence based on motion smoothness
            this._updateConfidence(this.position);
        }

        if (rotation) {
            this.quaternion.copy(rotation);
        }
    }

    /**
     * Updates only the grip state based on gestures.
     * @param {boolean} isClosedFist - True if the closed fist gesture is detected
     * @param {boolean} isOpenPalm - True if the open palm gesture is detected
     */
    updateGrip(isClosedFist, isOpenPalm) {
        const now = performance.now();

        // Track when we last saw closed fist
        if (isClosedFist) {
            this._lastClosedFistTime = now;
        }

        // Only turn on grip with closed fist, only turn off with open palm or timeout
        // Ignore button activation if confidence is too low
        if (this._buttonsBlocked) {
            // When blocked, treat as if no gesture detected (but don't force release)
            if (this.grip_Held && (now - this._lastClosedFistTime > this.gripTimeout)) {
                this.grip_Held = false;
            }
        } else if (isClosedFist) {
            this.grip_Held = true;
        } else if (isOpenPalm) {
            // Open palm releases all buttons
            this.grip_Held = false;
            this.trigger_Held = false;
            this.buttonA_Held = false;
            this.buttonB_Held = false;
        } else if (this.grip_Held && (now - this._lastClosedFistTime > this.gripTimeout)) {
            // Timeout: no closed_fist detected for gripTimeout ms while grip_Held
            this.grip_Held = false;
        }
        // Otherwise, grip_Held remains unchanged

        this.grip_Down = this.grip_Held && !this._wasGrip_Held;
        this._wasGrip_Held = this.grip_Held;
    }

    /**
     * Updates the trigger button state.
     * Trigger is sticky: only turns on with isHeld, turns off with isOpenPalm or isClosedFist.
     * @param {boolean} isHeld - True if the trigger gesture is detected
     * @param {boolean} isOpenPalm - True if open palm gesture is detected
     * @param {boolean} isClosedFist - True if closed fist gesture is detected
     */
    updateTrigger(isHeld, isOpenPalm, isClosedFist) {
        // Ignore activation when buttons are blocked due to low confidence
        if (!this._buttonsBlocked && isHeld) {
            this.trigger_Held = true;
        } else if (isOpenPalm || isClosedFist) {
            this.trigger_Held = false;
        }
        // Otherwise trigger_Held remains unchanged

        this.trigger_Down = this.trigger_Held && !this._wasTrigger_Held;
        this.trigger_Up = !this.trigger_Held && this._wasTrigger_Held;
        this._wasTrigger_Held = this.trigger_Held;
    }

    /**
     * Updates the button A state.
     * @param {boolean} isHeld - True if the button A gesture is detected
     */
    updateButtonA(isHeld) {
        // Ignore activation when buttons are blocked due to low confidence
        if (!this._buttonsBlocked) {
            this.buttonA_Held = isHeld;
        } else {
            this.buttonA_Held = false;
        }
        this.buttonA_Down = this.buttonA_Held && !this._wasButtonA_Held;
        this._wasButtonA_Held = this.buttonA_Held;
    }

    /**
     * Updates the button B state.
     * @param {boolean} isHeld - True if the button B gesture is detected
     */
    updateButtonB(isHeld) {
        // Ignore activation when buttons are blocked due to low confidence
        if (!this._buttonsBlocked) {
            this.buttonB_Held = isHeld;
        } else {
            this.buttonB_Held = false;
        }
        this.buttonB_Down = this.buttonB_Held && !this._wasButtonB_Held;
        this._wasButtonB_Held = this.buttonB_Held;
    }

    /**
     * Updates the button C state.
     * @param {boolean} isHeld - True if the button C gesture is detected
     */
    updateButtonC(isHeld) {
        // Ignore activation when buttons are blocked due to low confidence
        if (!this._buttonsBlocked) {
            this.buttonC_Held = isHeld;
        } else {
            this.buttonC_Held = false;
        }
        this.buttonC_Down = this.buttonC_Held && !this._wasButtonC_Held;
        this._wasButtonC_Held = this.buttonC_Held;
    }
}

File: openframeworks/nap-xtz-client/src/Controller.cpp
#include "Controller.h"

#include "NodeUtils.h"

#include <cmath>

//--------------------------------------------------------------
Controller::Controller() {
    init(KalmanSettings());
}

//--------------------------------------------------------------
Controller::Controller(const KalmanSettings & settings) {
    init(settings);
}

//--------------------------------------------------------------
void Controller::init(const KalmanSettings & settings) {
    kalmanEnabled = settings.enabled;

    // Drawing: half the smoothing, so the line follows the fingertip.
    const float drawQ = settings.Q * 2.0f;
    const float drawR = settings.R * 0.5f;
    drawKalmanX = KalmanFilter1D(drawQ, drawR);
    drawKalmanY = KalmanFilter1D(drawQ, drawR);
    drawKalmanZ = KalmanFilter1D(drawQ, drawR);

    // Navigation: double the smoothing. Grabbing the world with a jittery
    // position would shake the whole drawing, so this one lags on purpose.
    const float navQ = settings.Q * 0.5f;
    const float navR = settings.R * 2.0f;
    navKalmanX = KalmanFilter1D(navQ, navR);
    navKalmanY = KalmanFilter1D(navQ, navR);
    navKalmanZ = KalmanFilter1D(navQ, navR);
}

//--------------------------------------------------------------
void Controller::clearButtons() {
    grip_Down = grip_Held = false;
    trigger_Down = trigger_Held = trigger_Up = false;
    buttonA_Down = buttonA_Held = false;
    buttonB_Down = buttonB_Held = false;
    buttonC_Down = buttonC_Held = false;

    wasGrip_Held = false;
    wasTrigger_Held = false;
    wasButtonA_Held = false;
    wasButtonB_Held = false;
    wasButtonC_Held = false;
}

//--------------------------------------------------------------
void Controller::updatePose(const glm::vec3 & position) {
    if (kalmanEnabled) {
        drawPosition = glm::vec3(
            drawKalmanX.update(position.x),
            drawKalmanY.update(position.y),
            drawKalmanZ.update(position.z));

        navPosition = glm::vec3(
            navKalmanX.update(position.x),
            navKalmanY.update(position.y),
            navKalmanZ.update(position.z));

        // The node itself sits at the navigation position: it's what the
        // on-screen pointer and the palette hit test both follow.
        setPosition(navPosition);
    } else {
        setPosition(position);
        drawPosition = position;
        navPosition = position;
    }

    updateConfidence(getPosition());
}

//--------------------------------------------------------------
void Controller::updatePoseMissing() {
    // Deliberately empty. A hand that leaves the frame for a moment keeps its
    // last pose, so a stroke isn't yanked to the origin by one dropped frame.
}

//--------------------------------------------------------------
void Controller::updateConfidence(const glm::vec3 & currentPosition) {
    if (!hasHistory) {
        prevPosition = currentPosition;
        prevVelocity = glm::vec3(0.0f);
        hasHistory = true;
        return;
    }

    const glm::vec3 velocity = currentPosition - prevPosition;
    const glm::vec3 acceleration = velocity - prevVelocity;
    const float accelMag = glm::length(acceleration);

    // 0 acceleration -> 1.0, kAccelMidpoint -> ~0.5, and falling away from there.
    const float instantConfidence = std::exp(-accelMag / kAccelMidpoint * 0.693f); // ln(2)

    confidence = confidence * (1.0f - kConfidenceSmoothing) + instantConfidence * kConfidenceSmoothing;
    confidence = std::max(0.0f, std::min(1.0f, confidence));

    // Separate thresholds so a score hovering at the boundary doesn't flip the
    // gate on and off every frame.
    if (buttonsBlocked) {
        if (confidence > kUnblockThreshold) buttonsBlocked = false;
    } else {
        if (confidence < kBlockThreshold) buttonsBlocked = true;
    }

    prevPosition = currentPosition;
    prevVelocity = velocity;
}

//--------------------------------------------------------------
glm::vec3 Controller::getDrawPosition() const {
    if (getParent() == nullptr) return drawPosition;
    return NapDraw::localToWorld(*getParent(), drawPosition);
}

//--------------------------------------------------------------
glm::vec3 Controller::getNavPosition() const {
    if (getParent() == nullptr) return navPosition;
    return NapDraw::localToWorld(*getParent(), navPosition);
}

//--------------------------------------------------------------
// Grip latches: a fist closes it, an open palm opens it. Nothing else releases
// it except the timeout, so the hand can be turned or partly hidden mid-grab
// without dropping the world.
void Controller::updateGrip(bool isClosedFist, bool isOpenPalm) {
    const uint64_t now = ofGetElapsedTimeMillis();

    if (isClosedFist) lastClosedFistTime = now;

    if (buttonsBlocked) {
        // Tracking is poor: don't act on gestures, but still honour the timeout
        // so a grip can't get stuck on.
        if (grip_Held && (now - lastClosedFistTime > gripTimeout)) {
            grip_Held = false;
        }
    } else if (isClosedFist) {
        grip_Held = true;
    } else if (isOpenPalm) {
        // An open palm is the universal release.
        grip_Held = false;
        trigger_Held = false;
        buttonA_Held = false;
        buttonB_Held = false;
    } else if (grip_Held && (now - lastClosedFistTime > gripTimeout)) {
        grip_Held = false;
    }
    // Otherwise grip_Held is left as it was.

    grip_Down = grip_Held && !wasGrip_Held;
    wasGrip_Held = grip_Held;
}

//--------------------------------------------------------------
// The trigger latches the same way: Pointing_Up starts a stroke, and either an
// open palm or a fist ends it.
void Controller::updateTrigger(bool isHeld, bool isOpenPalm, bool isClosedFist) {
    if (!buttonsBlocked && isHeld) {
        trigger_Held = true;
    } else if (isOpenPalm || isClosedFist) {
        trigger_Held = false;
    }

    trigger_Down = trigger_Held && !wasTrigger_Held;
    trigger_Up = !trigger_Held && wasTrigger_Held;
    wasTrigger_Held = trigger_Held;
}

//--------------------------------------------------------------
void Controller::updateButtonA(bool isHeld) {
    buttonA_Held = buttonsBlocked ? false : isHeld;
    buttonA_Down = buttonA_Held && !wasButtonA_Held;
    wasButtonA_Held = buttonA_Held;
}

//--------------------------------------------------------------
void Controller::updateButtonB(bool isHeld) {
    buttonB_Held = buttonsBlocked ? false : isHeld;
    buttonB_Down = buttonB_Held && !wasButtonB_Held;
    wasButtonB_Held = buttonB_Held;
}

//--------------------------------------------------------------
void Controller::updateButtonC(bool isHeld) {
    buttonC_Held = buttonsBlocked ? false : isHeld;
    buttonC_Down = buttonC_Held && !wasButtonC_Held;
    wasButtonC_Held = buttonC_Held;
}
